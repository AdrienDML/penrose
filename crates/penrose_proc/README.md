## proc-macros for the Penrose tiling window manager library

Most (if not all) of the proc-macros in this crate are consumed internally in
the main [penrose][1] crate and are not intended for use outside of that
context. Please see the main penrose crate for further details.

  [1]: https://crates.io/crates/penrose
