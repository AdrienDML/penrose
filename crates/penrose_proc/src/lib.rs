//! Proc macros for use in the main Penrose crate
use proc_macro::TokenStream;

mod stub;
mod validate_bindings;

use stub::stubbed_companion_trait_inner;
use validate_bindings::validate_user_bindings_inner;

/// This is an internal macro that is used as part of `gen_keybindings` to validate user provided
/// key bindings at compile time using xmodmap.
///
/// It is not intended for use outside of that context and may be modified and updated without
/// announcing breaking API changes.
///
/// ```no_run
/// # use penrose_proc::validate_user_bindings;
/// validate_user_bindings!(
///     ( "M-a", "M-b" )
///     (
///         ( ( "M-{}", "M-S-{}" ) ( "1", "2", "3" ) )
///     )
/// );
/// ```
#[proc_macro]
pub fn validate_user_bindings(input: TokenStream) -> TokenStream {
    validate_user_bindings_inner(input)
}

#[proc_macro_attribute]
pub fn stubbed_companion_trait(args: TokenStream, input: TokenStream) -> TokenStream {
    stubbed_companion_trait_inner(args, input)
}
