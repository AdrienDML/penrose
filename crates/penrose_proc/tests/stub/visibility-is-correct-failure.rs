// The stub trait has the same visibility as the parent
mod inner {
    use penrose_proc::stubbed_companion_trait;

    #[stubbed_companion_trait]
    trait Foo {
        #[stub("red")]
        fn color(&self, x: u32) -> &str;
    }
}

struct MyStruct {}
// Should fail due to being private
impl inner::StubFoo for MyStruct {}

fn main() {}
