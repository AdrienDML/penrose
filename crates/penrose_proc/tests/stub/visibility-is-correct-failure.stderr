error[E0603]: trait `StubFoo` is private
  --> $DIR/visibility-is-correct-failure.rs:14:13
   |
14 | impl inner::StubFoo for MyStruct {}
   |             ^^^^^^^ private trait
   |
note: the trait `StubFoo` is defined here
  --> $DIR/visibility-is-correct-failure.rs:5:5
   |
5  |     #[stubbed_companion_trait]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
